### PR TITLE
List all nodes and occupy cidr map before starting allocations

### PR DESCRIPTION
**Pull request by bprashanth**
Manually tested by starting a 200 node cluster with frequent controller manager restarts.
Fixes https://github.com/kubernetes/kubernetes/issues/29058

### CODE REVIEW

File: pkg/controller/node/cidr_allocator.go [hunk 1]
```diff
@@ -82,6 +82,17 @@ func NewCIDRRangeAllocator(client clientset.Interface, clusterCIDR *net.IPNet, s
 	} else {
 		glog.V(0).Info("No Service CIDR provided. Skipping filtering out service addresses.")
 	}
+	for _, node := range nodeList.Items {
+		if node.Spec.PodCIDR == "" {
+			glog.Infof("Node %v has no CIDR, ignoring", node.Name)
```
Comment (justinsb) [inline pkg/controller/node/cidr_allocator.go]: Did you want to 'continue' here

Comment (justinsb) [inline pkg/controller/node/cidr_allocator.go]: Doesn't really matter either way, i guess😃

Comment (bprashanth) [inline pkg/controller/node/cidr_allocator.go]: yeah that's just for logging

**Comment by justinsb** [comment]
Haven't tested but looks right for a backport. I think there is still a theoretical race but think it is very unlikely now

**Comment by bprashanth** [comment]
I think as long as we do it in a STW phase we know we're not allocating any CIDRS, so the list has all our previous allocations. Please, please try to shoot holes through this :)

**Comment by bprashanth** [comment]
IF we have an inflith request stuck in apiserver memory that doesn't show up in our list but makes it to etcd after we populate the map, and in that time we allocate that cidr, things might go wrong. Apiserver should discard all broken connection writes though, and if it made it through to the datastore, it should show up in the list.

**Comment by bprashanth** [comment]
Cleaned it up a bit, deployed to 3 node cluster with restarting nodecontroller and it no dupes. Trying a larger cluster.

**Comment by justinsb** [comment]
The only failure case I could come up with was if there were no CIDRs left for the nodes, and we deleted and created a bunch of nodes in between the two lists, we might incorrectly deny some node allocations which we could satisfy.  But (1) the window is tiny and (2) the existing code doesn't handle the no-cidrs-left case very well anyway and (3) this isn't a double-allocation so not a disaster.  So I think this is great (presuming we don't spot anything), and has the advantage that it isn't an impossible cherry-pick onto the multiple branches.

**Comment by bgrant0607** [comment]
```
FAIL	k8s.io/kubernetes/contrib/mesos/cmd/km [build failed]
```
@mtaufen encountered this a couple days ago.

**Comment by bgrant0607** [comment]
I don't see an issue filed for that build failure. Checking whether it's a flake.
@k8s-bot unit test this: #IGNORE

**Comment by bprashanth** [comment]
contrib mesos has their own controller manager startup routine that's calling the same nodecontroller init function (contrib/pkg/controllermanager/controllermanager.go as opposed to kube-controller-manager). Invocation looks identical so I gave it the same failure mode.

**Comment by bgrant0607** [comment]
LGTM

This does point out that controllers need a way to determine that they are reading the most up-to-date state, without performing an extra write. I also encountered that when toying with a fix for the bugs in updateCIDRAllocation. I'll think about that problem.

In the future, CIDR allocation may be a good use case for multi-key transactions. The Node API could keep a secondary index of allocated CIDRs and could fail node updates that would double-assign the same CIDR.

**Pull request merged**